#! /bin/sh /usr/share/dpatch/dpatch-run
## 24_prefs_stacking_fix.dpatch by Daniel T Chen <crimsun@ubuntu.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix stacking in Preferences dialog, taken from upstream svn changeset 
## DP: 13795. Closes: Malone #31981.

@DPATCH@

--- trunk/modules/gui/wxwidgets/preferences.cpp
+++ trunk/modules/gui/wxwidgets/preferences.cpp
@@ -84,4 +84,6 @@
     wxWindow *p_parent;
     vlc_bool_t b_advanced;
+
+    wxPanel *p_current;
 
     wxTreeItemId root_item;
@@ -606,9 +608,9 @@
 
     /* Sort all this mess */
-    wxTreeItemIdValue cookie; 
+    wxTreeItemIdValue cookie;
     size_t i_child_index;
     wxTreeItemId capability_item = GetFirstChild( root_item, cookie);
     for( i_child_index = 0;
-         (capability_item.IsOk() && 
+         (capability_item.IsOk() &&
           //(i_child_index < GetChildrenCount( plugins_item, FALSE )));
           (i_child_index < GetChildrenCount( root_item, FALSE )));
@@ -625,4 +627,6 @@
     p_sizer->Add( this, 1, wxEXPAND | wxALL, 0 );
     p_sizer->Layout();
+
+    p_current = NULL;
 
     /* Update Tree Ctrl */
@@ -690,4 +694,5 @@
         p_sizer->Remove( config_data->panel );
 #endif
+        p_current = NULL;
     }
 
@@ -775,8 +780,48 @@
     ConfigTreeData *config_data = NULL;
 
-    if( event.GetOldItem() )
-        config_data = FindModuleConfig( (ConfigTreeData *)GetItemData(
-                                        event.GetOldItem() ) );
-    if( config_data && config_data->panel )
+    if( p_current )
+    {
+        p_current->Hide();
+#if (wxCHECK_VERSION(2,5,0))
+        p_sizer->Detach( p_current  );
+#else
+        p_sizer->Remove( p_current );
+#endif
+        p_current = NULL;
+    }
+
+    /* Don't use event.GetItem() because we also send fake events */
+    config_data = FindModuleConfig( (ConfigTreeData *)GetItemData(
+                                    GetSelection() ) );
+    if( config_data )
+    {
+        if( !config_data->panel )
+        {
+            /* The panel hasn't been created yet. Let's do it. */
+            config_data->panel =
+                new PrefsPanel( p_parent, p_intf, p_prefs_dialog,
+                                config_data );
+            config_data->panel->SwitchAdvanced( b_advanced );
+        }
+        else
+        {
+            config_data->panel->SwitchAdvanced( b_advanced );
+            config_data->panel->Show();
+        }
+
+        p_current = config_data->panel;
+
+        p_sizer->Add( config_data->panel, 3, wxEXPAND | wxALL, 0 );
+        p_sizer->Layout();
+    }
+}
+
+void PrefsTreeCtrl::OnAdvanced( wxCommandEvent& event )
+{
+    b_advanced = event.GetInt();
+
+    ConfigTreeData *config_data = !GetSelection() ? NULL :
+        FindModuleConfig( (ConfigTreeData *)GetItemData( GetSelection() ) );
+    if( config_data  )
     {
         config_data->panel->Hide();
@@ -786,44 +831,5 @@
         p_sizer->Remove( config_data->panel );
 #endif
-    }
-
-    /* Don't use event.GetItem() because we also send fake events */
-    config_data = FindModuleConfig( (ConfigTreeData *)GetItemData(
-                                    GetSelection() ) );
-    if( config_data )
-    {
-        if( !config_data->panel )
-        {
-            /* The panel hasn't been created yet. Let's do it. */
-            config_data->panel =
-                new PrefsPanel( p_parent, p_intf, p_prefs_dialog,
-                                config_data );
-            config_data->panel->SwitchAdvanced( b_advanced );
-        }
-        else
-        {
-            config_data->panel->SwitchAdvanced( b_advanced );
-            config_data->panel->Show();
-        }
-
-        p_sizer->Add( config_data->panel, 3, wxEXPAND | wxALL, 0 );
-        p_sizer->Layout();
-    }
-}
-
-void PrefsTreeCtrl::OnAdvanced( wxCommandEvent& event )
-{
-    b_advanced = event.GetInt();
-
-    ConfigTreeData *config_data = !GetSelection() ? NULL :
-        FindModuleConfig( (ConfigTreeData *)GetItemData( GetSelection() ) );
-    if( config_data  )
-    {
-        config_data->panel->Hide();
-#if (wxCHECK_VERSION(2,5,0))
-        p_sizer->Detach( config_data->panel );
-#else
-        p_sizer->Remove( config_data->panel );
-#endif
+        p_current = NULL;
     }
 
